src/Embeddings/VectorStores/Elasticsearch/ElasticsearchVectorStore.php

```diff
@@ -0,0 +1,203 @@
+                'hash' => $document->hash,
+                'chunkNumber' => $document->chunkNumber,
+            ],
+        ]);
+    }
+
+    /**
+     * @param  Document[]  $documents
+     *
+     * @throws Exception
+     */
+    public function addDocuments(array $documents, int $numberOfDocumentsPerRequest = 0): void
+    {
+        if ($documents === []) {
+            return;
+        }
+        if ($documents[0]->embedding === null) {
+            throw new Exception('document embedding must be set before adding a document');
+        }
+        $this->setVectorDimIfNotSet(count((array) $documents[0]->embedding));
+
+        $params = ['body' => []];
+        foreach ($documents as $document) {
+            $params['body'][] = [
+                'index' => [
+                    '_index' => $this->indexName,
+                ],
+            ];
+            $params['body'][] = [
+                'embedding' => $document->embedding,
+                'content' => $document->content,
+                'formattedContent' => $document->formattedContent ?? '',
+                'sourceType' => $document->sourceType,
+                'sourceName' => $document->sourceName,
+                'hash' => $document->hash,
+                'chunkNumber' => $document->chunkNumber,
+            ];
+
+        }
+        $this->client->bulk($params);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param  array{filter?: string|array<string, mixed>, num_candidates?: int}  $additionalArguments
+     * num_candidates is used to tune approximate kNN for speed or accuracy (see : https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#tune-approximate-knn-for-speed-accuracy)
+     */
+    public function similaritySearch(array $embedding, int $k = 4, array $additionalArguments = []): array
+    {
+        $numCandidates = max(50, $k * 4);
+        if (array_key_exists('num_candidates', $additionalArguments)) {
+            $numCandidates = $additionalArguments['num_candidates'];
+        }
+        $searchParams = [
+            'index' => $this->indexName,
+            'body' => [
+                'knn' => [
+                    'field' => 'embedding',
+                    'query_vector' => $embedding,
+                    'k' => $k,
+                    'num_candidates' => $numCandidates,
+                ],
+                'sort' => [
+                    '_score' => [
+                        'order' => 'desc',
+                    ],
+                ],
+            ],
+        ];
+        if (array_key_exists('filter', $additionalArguments)) {
+            $searchParams['body']['knn']['filter'] = $additionalArguments['filter'];
+        }
+        /** @var array{hits: array{hits: array{array{_source: array{embedding: float[], content: string, formattedContent: string, sourceType: string, sourceName: string, hash: string, chunkNumber: int}}}}} $rawResponse */
+        $rawResponse = $this->client->search($searchParams);
```

Comment:
Here we get an empty array for the first run, but it's fine on the second one (with no changes). 
The requests for creating the index and adding document return a 200 on the first run.
If we add a sleep(1) before this line, we don't get the error anymore on the first run ! Perhaps the document processing isn't finished when this request is played ?

I don't know what to do next to debug the situation, any advises ?

Reply:
I added a refresh: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-refresh.html